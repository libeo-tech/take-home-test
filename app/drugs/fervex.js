import {DrugModel} from "./models/drug";

export class Fervex extends DrugModel {
  constructor(expiresIn, benefit) {
    super(DrugModel.drugs.FERVEX, expiresIn, benefit)
  }

  updateBenefitValue(data) {
    try {
      DrugModel.validate(data);

      if (data.expiresIn <= 0) {
        data.benefit = DrugModel.minBenefitValue;
        return data;
      } else if (data.benefit < DrugModel.maxBenefitValue) {
        if (data.expiresIn <= 10 && data.expiresIn > 5) {
          data.benefit = data.benefit + 2;
        } else if (data.expiresIn <= 5 && data.expiresIn > 0) {
          data.benefit = data.benefit + 3;
        } else {
          data.benefit = data.benefit + 1;
        }
      }
      if (data.benefit > DrugModel.maxBenefitValue) data.benefit = DrugModel.maxBenefitValue;

      return data;
    } catch (error) {
      throw error;
    }
  }

  updateExpiresInValue(data) {
    try {
      DrugModel.validate(data);

      return DrugModel.updateExpiresInValue(data);
    } catch (error) {
      throw error;
    }
  }
}
