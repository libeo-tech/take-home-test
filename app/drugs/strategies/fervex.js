import {Drug} from "../drug";
import {DrugStrategy} from "./drug-strategy";

export class FervexStrategy extends DrugStrategy {
  constructor(name) {
    super();
    this.name = name;
    this.updateBenefitValue = this._updateBenefitValue;
    this.updateExpiresInValue = this._updateExpiresInValue;
  }

  _updateBenefitValue(data) {
    try {
      Drug.validate(data);

      if (data.expiresIn <= 0) {
        data.benefit = Drug.minBenefitValue;
        return data;
      } else if (data.benefit < Drug.maxBenefitValue) {
        if (data.expiresIn <= 10 && data.expiresIn > 5) {
          data.benefit = data.benefit + 2;
        } else if (data.expiresIn <= 5 && data.expiresIn > 0) {
          data.benefit = data.benefit + 3;
        } else {
          data.benefit = data.benefit + 1;
        }
      }
      if (data.benefit > Drug.maxBenefitValue) data.benefit = Drug.maxBenefitValue;

      return data;
    } catch (error) {
      throw error;
    }
  }

  _updateExpiresInValue(data) {
    try {
      Drug.validate(data);

      return DrugStrategy.updateExpiresInValue(data);
    } catch (error) {
      throw error;
    }
  }
}
