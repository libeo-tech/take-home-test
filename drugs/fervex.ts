import { Drug, DrugName } from "./drug";

export class Fervex extends Drug {
    constructor(expiresIn: number, benefit: number) {
        super(DrugName.FERVEX, expiresIn, benefit);
    }

    public updateDrug(): void {
        if (this.isExpired) {
            this.resetBenefit();
            return;
        }
        if (this.expiresIn <= 5) {
            this.decreaseExpiresIn();
            this.incrementBenefit(3);
            return;
        }
        if (this.expiresIn <= 10) {
            this.decreaseExpiresIn();
            this.incrementBenefit(2);
            return;
        }
    }
}