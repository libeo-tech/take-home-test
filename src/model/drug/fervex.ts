import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class Fervex extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.FERVEX, expiresIn, benefit);
  }

  public updateValues() {
    if (this.expiresIn > 10) {
      this.benefit += 1;
    } else if (this.expiresIn > 5) {
      this.benefit += 2;
    } else if (this.expiresIn > 0) {
      this.benefit += 3;
    } else {
      this.benefit = 0;
    }
    this.expiresIn -= 1;

    return this;
  }
}
