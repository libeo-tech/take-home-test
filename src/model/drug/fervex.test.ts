import { Fervex } from './fervex';

describe('Fervex', () => {
  let fervex: Fervex;

  beforeEach(() => {
    fervex = new Fervex(30, 5);
  });

  it('should not allow for a negative benefit', () => {
    fervex.benefit = -5;
    expect(fervex).toMatchSnapshot();
  });

  it('should not allow for a benefit greater than 50', () => {
    fervex.benefit = 51;
    expect(fervex).toMatchSnapshot();
  });

  it('should return expiration status', () => {
    expect(fervex.isExpired()).toBe(false);
    fervex.expiresIn = 0;
    expect(fervex.isExpired()).toBe(true);
  });

  it('should increase benefit with time', () => {
    fervex.updateValues();
    expect(fervex).toMatchSnapshot();
  });

  test('benefit + 2 if 10 >= expiresIn > 5', () => {
    fervex.expiresIn = 10;
    fervex.updateValues();
    expect(fervex).toMatchSnapshot();

    fervex.expiresIn = 6;
    fervex.updateValues();
    expect(fervex).toMatchSnapshot();
  });

  test('benefit + 3 if 5 >= expiresIn > 0', () => {
    fervex.expiresIn = 5;
    fervex.updateValues();
    expect(fervex).toMatchSnapshot();

    fervex.expiresIn = 1;
    fervex.updateValues();
    expect(fervex).toMatchSnapshot();
  });

  test('benefit drops to 0 if 0 >= expiresIn > -âˆž', () => {
    fervex.expiresIn = 0;
    fervex.updateValues();
    expect(fervex).toMatchSnapshot();
  });

  it('should be serializable to JSON', () => {
    expect(fervex.toJSON()).toMatchSnapshot();
  });

  it('should be serializable to string', () => {
    expect(fervex.toString()).toMatchSnapshot();
  });
});
