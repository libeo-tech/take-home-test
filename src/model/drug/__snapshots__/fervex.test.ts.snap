// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fervex benefit + 2 if 10 >= expiresIn > 5 1`] = `
Object {
  "benefit": 7,
  "expiresIn": 9,
  "name": "Fervex",
}
`;

exports[`Fervex benefit + 2 if 10 >= expiresIn > 5 2`] = `
Object {
  "benefit": 9,
  "expiresIn": 5,
  "name": "Fervex",
}
`;

exports[`Fervex benefit + 3 if 5 >= expiresIn > 0 1`] = `
Object {
  "benefit": 8,
  "expiresIn": 4,
  "name": "Fervex",
}
`;

exports[`Fervex benefit + 3 if 5 >= expiresIn > 0 2`] = `
Object {
  "benefit": 11,
  "expiresIn": 0,
  "name": "Fervex",
}
`;

exports[`Fervex benefit drops to 0 if 0 >= expiresIn > -âˆž 1`] = `
Object {
  "benefit": 0,
  "expiresIn": -1,
  "name": "Fervex",
}
`;

exports[`Fervex should be serializable to JSON 1`] = `
Object {
  "benefit": 5,
  "expiresIn": 30,
  "name": "Fervex",
}
`;

exports[`Fervex should be serializable to string 1`] = `"{\\"name\\":\\"Fervex\\",\\"expiresIn\\":30,\\"benefit\\":5}"`;

exports[`Fervex should increase benefit with time 1`] = `
Object {
  "benefit": 6,
  "expiresIn": 29,
  "name": "Fervex",
}
`;

exports[`Fervex should not allow for a benefit greater than 50 1`] = `
Object {
  "benefit": 50,
  "expiresIn": 30,
  "name": "Fervex",
}
`;

exports[`Fervex should not allow for a negative benefit 1`] = `
Object {
  "benefit": 0,
  "expiresIn": 30,
  "name": "Fervex",
}
`;
